### ISSUE TITLE
Custom dataset training resampling several times during mAP validation

### ISSUE THREAD

**Issue by jdonovanCS** [dbolya/yolact#432]
@dbolya, I have around 1800 images, with around 7061 annotations (currently each annotation is separate for each instance of class, even if it is the same class). Initially, most of my images were very large so I resized all of them to have a larger dimension of 640 and to keep their ratio of h to w. I've created a coco annotation json file with masks, bboxes, categories, etc. for each image-instance pairing. I have edited the config with my classes:

`MINC_LABEL_MAP = {0: 1, 1: 2, 2: 3, 3: 4, 4: 5, 5: 6, 6: 7, 7: 8, 8: 9, 9: 10,
                    10: 11, 11: 12, 12: 13, 13: 14, 14: 15, 15: 16, 16: 17, 17: 18, 18: 19, 19: 20,
                    20: 21, 21: 22, 22: 23}

MINC_CLASSES = ("brick","carpet","ceramic","fabric","foliage","food",
"glass","hair","leather","metal","mirror","other","painted","paper","plastic",
"polishedstone","skin","sky","stone","tile","wallpaper","water","wood")`

as well as to point to these classes:

`yolact_base_config = coco_base_config.copy({
    'name': 'yolact_base',

    # Dataset stuff
    'dataset': minc_dataset,
    'num_classes': len(minc_dataset.class_names) + 1,

    # Image Size
    'max_size': 550,`

and

`minc_dataset = dataset_base.copy({
    'name': 'MINC',

    'train_images': './data/coco/images-minc',
    'train_info': './data/minc/instances_all.json',

    'Valid_images': './data/coco/images-minc',
    'valid_info': './data/minc/instances_all.json',

    'has_gt': True,

    'label_map': MINC_LABEL_MAP,
    'class_names': MINC_CLASSES
})`

I've been training for about 1.5 days using --config=yolact_base_config --batch_size=5 and resuming each time I have to stop it.

My question is two-fold..

1. I am getting a lot of output when computing validation mAP such as below which seems wrong.

`Computing validation mAP (this may take a while)...

Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
Warning: Augmentation output an example with no ground truth. Resampling...
`

2. I am getting very low numbers so far on the actual mAP such as below. I've included a line of the training output, so that you can see those numbers as well. Since that may make a difference in how you answer this (ie. It may just be too early in the training to know much).

Calculating mAP...

       |  all  |  .50  |  .55  |  .60  |  .65  |  .70  |  .75  |  .80  |  .85  |  .90  |  .95  |
-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+
   box |  0.04 |  0.20 |  0.13 |  0.08 |  0.01 |  0.00 |  0.00 |  0.00 |  0.00 |  0.00 |  0.00 |
  mask |  0.01 |  0.03 |  0.02 |  0.00 |  0.00 |  0.00 |  0.00 |  0.00 |  0.00 |  0.00 |  0.00 |
-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+-------+

[103]   36980 || B: 5.843 | C: 4.773 | M: 90.437 | S: 0.340 | T: 101.392 || ETA: 26 days, 2:01:11 || timer: 0.794
[103]   36990 || B: 5.724 | C: 4.789 | M: 90.071 | S: 0.329 | T: 100.913 || ETA: 26 days, 1:32:27 || timer: 0.741

If this does not pertain to the above question or if it is too much, feel free to open a new issue or contact me about doing so.

Thank you!

_Originally posted by @jdonovanCS in https://github.com/dbolya/yolact/issues/403#issuecomment-618698324_

**Comment by ic** [comment]
No solution here yet, just some more information "down the road" ahead.

I have similar settings here, slightly more images and annotations (3000 and 9000), and similar observation. The custom data set is also structured the same way, based on the COCO format, without using the format features to gather annotations per file (so, as here, one annotation per object instance). This last decision is based on the [Yolact training code for COCO](https://github.com/dbolya/yolact/blob/master/data/coco.py#L115).

On this trial, I have gone to:

    [1370]  222100 ||...

Where the numbers are 10-20x higher on the box mAP, but still very low on the mask mAP.

Trying `eval.py` with the best model shows the performance on boxes is taking off, but no mask at this point (unsurprisingly).

On the warnings related to augmentation without GT, I had them at first and solved them by fixing the format of my COCO custom manifest. Sorry, I did not keep track of what it was exactly---I just remember it was a problem between my manifest and the collecting code in Yolact.

**Comment by jdonovanCS** [comment]
So I figured out part of my issue by using the [COCO Image Viewer Notebook](https://github.com/Tony607/voc2coco/blob/master/COCO_Image_Viewer.ipynb). It allowed me to see that my COCO dataset that I had created had some issues with lining up the bounding boxes and masks with the actual objects in the image. I have fixed that now and am running YOLACT again on this dataset. So far, I am not getting the resampling issue. The mAP at this point is still pretty low, but it has only been a couple of hours.

**Comment by ic** [comment]
So it was a COCO-encoding issue on your side too... May the couple hours turn well, and successfully close this issue!

Going through the issue again, I now see that your first problem was a sign the dataset was broken. I have now read most of the source, and the `has_gt=True` in your configuration lets the script believe ground truth information is available (that is, the bounding box and mask data corresponding to each image). The warning seems fine for an occasional missing GT, but it seems it was too frequent in your initial run.

> Baseline: If one see too many such warnings, something odd's happening.

---

On my last comment, I mentioned training history plots over epochs. I have run through several errors before getting expected results, and can say now that both box and mask mAPs increase visibly right from the first few epochs. No plot, but the understanding that if you do not see both mAPs rising, something's at odd.

To be clear, I have a run on an old Nvidia 1050 mobile (4Gb), where batch size cannot exceed 1 (with default 550x550 settings), and even on this one the mAPs increase before epoch 10 (last run was epoch 7 for masks, and before for boxes).

**Comment by harishkool** [comment]
One of the other reasons for getting those warnings are `discard_box_width` and `discard_box_height` in the config file. By default the value is `4 / 550`, you can change that if your dataset has very small objects and can get rid of those warnings.

**Comment by jdonovanCS** [comment]
I'm closing this issue as I have found the solution as mentioned above. Others have also provided input that will help with this issue.

**Comment by Isfate** [comment]
@jdonovanCS When I was training Yolact on a custom data set, I ran into the same problem as you. I would like to ask you about the solution to this problem. Looking forward to your reply.

**Comment by surefyyq** [comment]
> One of the other reasons for getting those warnings are `discard_box_width` and `discard_box_height` in the config file. By default the value is `4 / 550`, you can change that if your dataset has very small objects and can get rid of those warnings.

So may i ask what's the logic of 4/550, since 550 is the default max_size, does it mean quarter of 550？Should that be 550/4?
